jetty-core/jetty-session/src/main/java/org/eclipse/jetty/session/AbstractSessionManager.java

```diff
@@ -1178,103 +1200,107 @@ protected RequestedSession resolveRequestedSessionId(Request request)
-                    LOG.debug("Got Session ID {} from URL", requestedSessionId);
+        if (ids == null)
+            return NO_REQUESTED_SESSION;
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("Got Session IDs {} {}", ids, cookieIds);
+
+        for (int i = 0; i < ids.size(); i++)
+        {
+            String id = ids.get(i);
+            if (session == null)
+            {
+                //we currently do not have a session selected, use this one if it is valid
+                ManagedSession s = getManagedSession(id);
+                if (s != null && s.isValid())
+                {
+                    //associate it with the request so its reference count is decremented as the
+                    //request exits
+                    requestedSessionId = id;
+                    requestedSessionIdFromCookie = i < cookieIds;
+                    session = s;
 
-                session = getManagedSession(requestedSessionId);
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Selected session {}", session);
+                }
+                else
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("No session found for session cookie id {}", id);
+
+                    //if we don't have a valid session id yet, just choose the first one
+                    if (requestedSessionId == null)
+                    {
+                        requestedSessionId = id;
+                        requestedSessionIdFromCookie = i < cookieIds;
+                    }
+                }
+            }
+            else
+            {
+                //we currently have a valid session selected. We will throw an error
+                //if there is a _different_ valid session id cookie. Duplicate ids, or
+                //invalid session ids are ignored
+                if (!session.getId().equals(getSessionIdManager().getId(id)))
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Multiple different valid session ids: {}, {}", requestedSessionId, id);
+
+                    //load the session to see if it is valid or not
+                    ManagedSession s = getManagedSession(id);
+                    if (s != null && s.isValid())
+                    {
+                        //release both sessions straight away??
+                        try
+                        {
+                            _sessionCache.release(session);
+                            _sessionCache.release(s);
+                        }
+                        catch (Exception x)
+                        {
+                            if (LOG.isDebugEnabled())
+                                LOG.debug("Error releasing duplicate valid session: {}", id);
+                        }
+
+                        throw new BadMessageException("Duplicate valid session cookies: " + requestedSessionId + " ," + id);
```

Comment:
```suggestion
                        throw new BadMessageException("Duplicate valid sessions: " + requestedSessionId + " ," + id);
```

We should also output the source  of  the duplicates: could be one cookie and  one parameter, or multiple  cookies or multiple parameters